From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Thu, 16 Apr 2026 00:06:45 +0200
Subject: st/focus-manager: Allow to get the group from a ClutterActor

A focus group can also contain a simple ClutterActor and in fact this is
the case when the SearchController is, where a ClutterText owns the
focus.

Since the focus manager group getter was already using the ClutterActor
API there is no need to restrict this to StWidget's only.

Origin: https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/4181
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/gnome-shell/+bug/2148587
---
 src/st/st-focus-manager.c | 10 ++++------
 src/st/st-focus-manager.h |  5 ++++-
 2 files changed, 8 insertions(+), 7 deletions(-)

diff --git a/src/st/st-focus-manager.c b/src/st/st-focus-manager.c
index cb1d928..99c3a96 100644
--- a/src/st/st-focus-manager.c
+++ b/src/st/st-focus-manager.c
@@ -257,20 +257,18 @@ st_focus_manager_remove_group (StFocusManager *manager,
 /**
  * st_focus_manager_get_group:
  * @manager: the #StFocusManager
- * @widget: an #StWidget
+ * @actor: a #ClutterActor
  *
- * Checks if @widget is inside a focus group, and if so, returns
+ * Checks if @actor is inside a focus group, and if so, returns
  * the root of that group.
  *
  * Returns: (transfer none): the focus group root, or %NULL if
- * @widget is not in a focus group
+ * @actor is not in a focus group
  */
 StWidget *
 st_focus_manager_get_group (StFocusManager *manager,
-                            StWidget       *widget)
+                            ClutterActor   *actor)
 {
-  ClutterActor *actor = CLUTTER_ACTOR (widget);
-
   while (actor && !g_hash_table_lookup (manager->groups, actor))
     actor = clutter_actor_get_parent (actor);
 
diff --git a/src/st/st-focus-manager.h b/src/st/st-focus-manager.h
index 7318344..e7f16cf 100644
--- a/src/st/st-focus-manager.h
+++ b/src/st/st-focus-manager.h
@@ -41,6 +41,9 @@ void            st_focus_manager_add_group     (StFocusManager *manager,
 void            st_focus_manager_remove_group  (StFocusManager *manager,
                                                 StWidget       *root);
 StWidget       *st_focus_manager_get_group     (StFocusManager *manager,
-                                                StWidget       *widget);
+                                                ClutterActor   *actor);
+
+gboolean        st_focus_manager_navigate_from_event (StFocusManager *manager,
+                                                      ClutterEvent   *event);
 
 G_END_DECLS
